<img src='images/logo_full.png'/>

## <p style="text-align: center;"> Python for Data Science </p>
The course is offered by [Ai Adventures](www.aiadventures.in). The notebooks are created by [Pranav Uikey]() and [Ankur Singh](). This material is subject to the terms and conditions of the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license. Any use for commercial purpose is strictly prohibited.

# Loops

## *while* Loop Statements

You can make a block of code execute over and over again with a *while* statement. The code in a *while* clause will be executed as long as the *while* statement’s condition is True. In code, a *while* statement always consists of the following:

1. The *while* keyword

2. A condition (that is, an expression that evaluates to True or False)

3. A colon

4. Starting on the next line, an indented block of code (called the while clause)

You can see that a while statement looks similar to an *if* statement. The difference is in **how they behave**. At the end of an *if* clause, the program execution continues after the *if* statement. But at the end of a *while* clause, the program execution jumps back to the start of the *while* statement. The *while* clause is often called the **while loop** or just the **loop**.

Let’s look at an *if* statement and a *while* loop that use the same condition and take the same actions based on that condition.

In [1]:
# Here is the code with an if statement:

spam = 0
if spam < 5:
    print('Hello, world.')
    spam = spam + 1

Hello, world.


In [2]:
# Here is the code with a while statement:

spam = 0
while spam < 5:
    print('Hello, world.')
    spam = spam + 1


Hello, world.
Hello, world.
Hello, world.
Hello, world.
Hello, world.


These statements are similar—both *if* and *while* check the value of spam, and if it’s less than five, they print a message. But when you run these two code snippets, something very different happens for each one. For the *if* statement, the output is simply "Hello, world.". But for the *while* statement, it’s "Hello, world." repeated five times! Take a look at the flowcharts for these two pieces of code, to see why this happens.

### <span style="color:brown">*if* statement:</span>
![](images/000091.png)

### <span style="color:brown">*while* statement:</span>
![](images/000094.png)

The code with the *if* statement checks the condition, and it prints Hello, world. only once if that condition is true. The code with the *while* loop, on the other hand, will print it five times. It stops after five prints because the integer in spam is incremented by one at the end of each loop iteration, which means that the loop will execute five times before *spam < 5* is False.

In the *while* loop, the condition is always checked at the start of each iteration (that is, each time the loop is executed). If the condition is True, then the clause is executed, and afterward, the condition is checked again. The first time the condition is found to be False, the *while* clause is skipped.

### An Annoying while Loop
Here’s a small example program that will keep asking you to type, literally, **your name**.

In [3]:
name = '' # (1)
while name != 'your name': # (2)
    print('Please type your name')
    name = input() # (3)
print('thank you!') # (4)

Please type your name


 y


Please type your name


 ai


Please type your name


 your name


thank you!


First, the program sets the name variable ❶ to an empty string. This is so that the name != 'your name' condition will evaluate to True and the program execution will enter the while loop’s clause ❷.

The code inside this clause asks the user to type their name, which is assigned to the name variable ❸. Since this is the last line of the block, the execution moves back to the start of the while loop and reevaluates the condition. If the value in name is not equal to the string 'your name', then the condition is True, and the execution enters the while clause again.

But once the user types your name, the condition of the while loop will be 'your name' != 'your name', which evaluates to False. The condition is now False, and instead of the program execution reentering the while loop’s clause, it skips past it and continues running the rest of the program ❹

If you never enter your name, then the while loop’s condition will never be False, and the program will just keep asking forever. Here, the input() call lets the user enter the right string to make the program move on. In other programs, the condition might never actually change, and that can be a problem. Let’s look at how you can break out of a while loop.

## *for* Loops and the *range()* Function

The *while* loop keeps looping while its condition is True (which is the reason for its name), but what if you want to execute a block of code only a certain number of times? You can do this with a *for* loop statement and the range() function.

In code, a *for* statement looks something like *for i in range(5)*: and always includes the following:

1. The *for* keyword

2. A variable name

3. The *in* keyword

4. A call to the *range()* method with up to three integers passed to it

5. A colon

6. Starting on the next line, an indented block of code (called the *for* clause)

In [4]:
print('My name is')
for i in range(5):
    print('Jimmy Five Times (' + str(i) + ')')

My name is
Jimmy Five Times (0)
Jimmy Five Times (1)
Jimmy Five Times (2)
Jimmy Five Times (3)
Jimmy Five Times (4)


The code in the *for* loop’s clause is run five times. The first time it is run, the variable i is set to 0. The *print()* call in the clause will print Jimmy Five Times (0). After Python finishes an iteration through all the code inside the for loop’s clause, the execution goes back to the top of the loop, and the *for* statement increments *i* by one. This is why *range(5)* results in five iterations through the clause, with *i* being set to 0, then 1, then 2, then 3, and then 4. The variable *i* will go up to, but will not include, the integer passed to *range()*

#### <span style="color:brown">*for* statement</span>
![](images/000102.png)

##### An Equivalent *while* Loop
You can actually use a *while* loop to do the same thing as a *for* loop; *for* loops are just more concise and readable. Let’s rewrite it, to use a *while* loop equivalent of a *for* loop.

In [7]:
print('My name is')
i = 0
while i < 5:
    print('Jimmy Five Times (' + str(i) + ')')
    i = i+1

My name is
Jimmy Five Times (0)
Jimmy Five Times (1)
Jimmy Five Times (2)
Jimmy Five Times (3)
Jimmy Five Times (4)


As another *for* loop example, consider this story about the mathematician Karl Friedrich Gauss. When Gauss was a boy, a teacher wanted to give the class some busywork. The teacher told them to add up all the numbers from 0 to 100. Young Gauss came up with a clever trick to figure out the answer in a few seconds, but you can write a Python program with a *for* loop to do this calculation for you.

In [5]:
total = 0 # (1)
for num in range(101): # (2)
    total = total + num # (3)
print(total) # (4)

5050


The result should be 5,050. When the program first starts, the total variable is set to 0 ❶. The for loop ❷ then executes total = total + num ❸ 100 times. By the time the loop has finished all of its 100 iterations, every integer from 0 to 100 will have been added to total. At this point, total is printed to the screen ❹. Even on the slowest computers, this program takes less than a second to complete.

### range() function

#### The Starting, Stopping, and Stepping Arguments to range()
Some functions can be called with multiple arguments separated by a comma, and *range()* is one of them. This lets you change the integer passed to *range()* to follow any sequence of integers, including starting at a number other than zero.

In [8]:
for i in range(12, 16):
    print(i)

12
13
14
15


The first argument will be where the *for* loop’s variable starts, and the second argument will be up to, but not including, the number to stop at.

The *range()* function can also be called with **three arguments**. The first **two arguments** will be the **start** and **stop values**, and the **third** will be the **step argument**. The **step** is the amount that the variable is increased by after each iteration.

In [9]:
for i in range(0, 10, 2):
    print(i)

0
2
4
6
8


So calling range(0, 10, 2) will count from zero to eight by intervals of two.

The range() function is flexible in the sequence of numbers it produces for for loops. For example, you can even use a **negative number** for the step argument to make the for loop count down instead of up.

In [10]:
for i in range(5, -1, -1):
    print(i)

5
4
3
2
1
0


### *for* loops work really well with collections

#### Using *for* Loops with Lists

You learned about using for loops to execute a block of code a certain number of times. Technically, a for loop repeats the code block once for each value in a list or list-like value. For example, if you ran this code:

In [51]:
for i in range(4):
    print(i)

0
1
2
3


This is because the return value from range(4) is a list-like value that Python considers similar to [0, 1, 2, 3]. The following program has the same output as the previous one:

In [52]:
for i in [0, 1, 2, 3]:
    print(i)

0
1
2
3


What the previous for loop actually does is loop through its clause with the variable i set to a successive value in the [0, 1, 2, 3] list in each iteration.

A common Python technique is to use range(len(someList)) with a for loop to iterate over the indexes of a list. 

In [53]:
supplies = ['pens', 'staplers', 'flame-throwers', 'binders']
for i in range(len(supplies)):
    print('Index ' + str(i) + ' in supplies is: ' + supplies[i])

Index 0 in supplies is: pens
Index 1 in supplies is: staplers
Index 2 in supplies is: flame-throwers
Index 3 in supplies is: binders


Using range(len(supplies)) in the previously shown for loop is handy because the code in the loop can access the index (as the variable i) and the value at that index (as supplies[i]). Best of all, range(len(supplies)) will iterate through all the indexes of supplies, no matter how many items it contains.

### Using *for* loops with Dictionaries
##### The *keys()*, *values()*, and *items()* Methods

There are three dictionary methods that will return list-like values of the dictionary’s keys, values, or both keys and values: keys(), values(), and items(). The values returned by these methods are not true lists: They cannot be modified and do not have an append() method. But these data types (dict_keys, dict_values, and dict_items, respectively) can be used in for loops. Lets see how these methods work,

In [115]:
spam = {'color': 'red', 'age': 42}

In [116]:
for v in spam.values():
    print(v)

red
42


Here, a for loop iterates over each of the values in the spam dictionary. A for loop can also iterate over the keys or both keys and values:

In [117]:
for k in spam.keys():
        print(k)

color
age


In [118]:
for i in spam.items():
        print(i)

('color', 'red')
('age', 42)


Using the keys(), values(), and items() methods, a for loop can iterate over the keys, values, or key-value pairs in a dictionary, respectively. Notice that the values in the dict_items value returned by the items() method are tuples of the key and value.

If you want a true list from one of these methods, pass its list-like return value to the list() function.

In [119]:
spam = {'color': 'red', 'age': 42}

In [120]:
spam.keys()

dict_keys(['color', 'age'])

In [121]:
list(spam.keys())

['color', 'age']

The list(spam.keys()) line takes the dict_keys value returned from keys() and passes it to list(), which then returns a list value of ['color', 'age'].

You can also use the **multiple assignment** trick in a for loop to assign the key and value to separate variables.

In [122]:
for k, v in spam.items():
        print('Key: ' + k + ' Value: ' + str(v))

Key: color Value: red
Key: age Value: 42


___
#### Example 
Here’s a program that uses a dictionary that contains other dictionaries in order to see who is bringing what to a picnic. The *totalBrought()* function can read this data structure and calculate the total number of an item being brought by all the guests.

In [129]:
allGuests = {'Alice': {'apples': 5, 'pretzels': 12},
                'Bob': {'ham sandwiches': 3, 'apples': 2},
                'Carol': {'cups': 3, 'apple pies': 1}}

In [130]:
def totalBrought(guests, item):
    numBrought = 0
    for k, v in guests.items(): # (1)
        numBrought = numBrought + v.get(item, 0) # (2)
        return numBrought

In [131]:
print('Number of things being brought:')
print(' - Apples         ' + str(totalBrought(allGuests, 'apples')))
print(' - Cups           ' + str(totalBrought(allGuests, 'cups')))
print(' - Cakes          ' + str(totalBrought(allGuests, 'cakes')))
print(' - Ham Sandwiches ' + str(totalBrought(allGuests, 'ham sandwiches')))
print(' - Apple Pies     ' + str(totalBrought(allGuests, 'apple pies')))

Number of things being brought:
 - Apples         5
 - Cups           0
 - Cakes          0
 - Ham Sandwiches 0
 - Apple Pies     0


Inside the totalBrought() function, the for loop iterates over the key-value pairs in guests ❶. Inside the loop, the string of the guest’s name is assigned to k, and the dictionary of picnic items they’re bringing is assigned to v. If the item parameter exists as a key in this dictionary, it’s value (the quantity) is added to numBrought ❷. If it does not exist as a key, the get() method returns 0 to be added to numBrought.

This may seem like such a simple thing to model that you wouldn’t need to bother with writing a program to do it. But realize that this same totalBrought() function could easily handle a dictionary that contains thousands of guests, each bringing thousands of different picnic items. Then having this information in a data structure along with the totalBrought() function would save you a lot of time!

You can model things with data structures in whatever way you like, as long as the rest of the code in your program can work with the data model correctly. When you first begin programming, don’t worry so much about the “right” way to model data. As you gain more experience, you may come up with more efficient models, but the important thing is that the data model works for your program’s needs.

### Summary

You can execute code over and over again in a loop while a certain condition evaluates to True.

When working with *while* loops pay special attention to **update** statements. Else, you will get trapped in an infinite *while* loop which is not a good thing to happen (trust me!)